### PR TITLE
add resource owner to permissions

### PR DESCRIPTION
**Pull request by mattlub**
update:

related #134 

- [x] make middleware
  - changes name from `rolePermission` to just `permissions`
  - adds functionality to check if the requesting user owns the resource
- [x] write tests
  - amend old middleware tests to use new syntax
  - write new tests in `permissions.test.js`

**Note:** The middleware has not been properly added to routes yet, when it is, full integration tests for the middleware will be created from the old controller tests
---
the auth is completetly separate from the JWT stuff
WIP: currently failing tests because auth is req'd
relates #121

### CODE REVIEW

**Review by m4v15 — CHANGES_REQUESTED** [review]
Get resource owner from the path?

**Comment by mattlub** [comment]
@m4v15 made the change

**Review by m4v15 — APPROVED** [review]

File: src/middleware/permissions.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  const pathName = url.parse(req.url).pathname
+  return routeResourceMapping[pathName.split('/')[1]]
+}
+
+module.exports =
+  ({ minSufficientRole, owningResourceIsSufficient }) => (req, res, next) => {
+    const resourceType = getResourceType(req)
+
+    if (owningResourceIsSufficient && !resourceType) {
+      next(boom.badImplementation())
```
Comment (des-des) [inline src/middleware/permissions.js]: :ok_hand:

File: src/middleware/permissions.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    if (!req.user || !req.user.id) {
+      return next(boom.unauthorized(auth.UNAUTHORIZED))
+    } else if (hasSufficientRole({ minSufficientRole })(req.user)) {
+      return next()
+    } else if (!owningResourceIsSufficient) {
+      return next(boom.unauthorized(auth.UNAUTHORIZED))
+    } else {
```
Comment (des-des) [inline src/middleware/permissions.js]: Okay am confused.
hasSufficientRole && (user is trying to modify resources that they do not own)

then we let them through?

this if / else branching not sure about.
```js
if (minRole && !hasSufficientRole(req, minRole)) { Boom.unauthorised }
 
if (needOwner && !hasSufficientOwnership(req) { Boom.unauthorised }
```
Either this is broken, or the naming is incorrect

File: tests/fixtures/events.json [hunk 3]
```diff
@@ -5,5 +5,18 @@
    "invalidEvent1": {},
    "invalidEvent2": { "en":{"name": "Dinner at Alreda"}, "categories": ["eating"] },
    "invalidEvent3": { "en":{"name": "Dinner at Alreda"}, "categories": [null, "dining"] },
-   "invalidEvent4": { "en":{"name": "Dinner at Alreda"}, "categories": []}
+   "invalidEvent4": { "en":{"name": "Dinner at Alreda"}, "categories": []},
+
+   "event": {
```
Comment (des-des) [inline tests/fixtures/events.json]: why is this called `event`. The ones above are called `validEvent`

File: tests/fixtures/users.json [hunk 4]
```diff
@@ -17,11 +18,11 @@
...
 		}
 	},
-	"regUser":{
-			"username": "m4v15",
-			"password": "hashthis",
-			"confirm-password":"hashthis",
-			"email": "newemail",
-			"name_en":"mavis"
+	"regUser": {
+		"username": "m4v15",
```
Comment (des-des) [inline tests/fixtures/users.json]: tabs??

File: tests/middleware/permissions.test.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+tape('checkUserOwnsResource with resource \'User\', and the correct user', t => {
+  checkUserOwnsResource(User)(user.id)(user)
+  .then(() => t.end())
+  .catch(err => t.end(err))
+})
+
+tape('checkUserOwnsResource with resource \'User\', and an incorrect user', t => {
+  checkUserOwnsResource(User)(user.id)(wrongUser)
+  .then(() => t.end('promise should reject, as user is wrong'))
+  .catch(() => t.end())
```
Comment (des-des) [inline tests/middleware/permissions.test.js:43]: Having a test without an assertion is kinda weird.
```js
t.ok(true, 'a good thing happened')
```
works

**Review by des-des — CHANGES_REQUESTED** [review]

File: tests/fixtures/events.json (hunk 3 above)
Comment (mattlub) [reply tests/fixtures/events.json]: removed, and used validEvent1

File: tests/fixtures/users.json (hunk 4 above)
Comment (mattlub) [reply tests/fixtures/users.json]: json formatted

File: tests/middleware/permissions.test.js (hunk 5 above)
Comment (mattlub) [reply tests/middleware/permissions.test.js:43]: I disagree, we run the tests to check there are no failures, it seems pointless to stick an extra `t.pass(...)` or `t.ok(true,...)` in there

File: src/middleware/permissions.js (hunk 2 above)
Comment (mattlub) [reply src/middleware/permissions.js]: minSufficientRole is the minimum role which would guarantee you access, regardless of whether the resource owner can also access.

so I did
```js
if (hasSufficientRole) { next }
if (!hasSufficientRole && hasSufficientOwnership) { next }
if (!hasSufficientRole && !hasSufficientOwnership) { boom(unauthorised) }
```

I'm not convinced what I've done will turn out to be the best method, but I'm not entirely sure what you're suggesting.

**Comment by mattlub** [comment]
@des-des I've changed the names now.

File: src/middleware/permissions.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    if (ownerIsPermitted && !resourceType) {
+      next(boom.badImplementation())
+    }
+
+    if (!req.user || !req.user.id) {
+      return next(boom.unauthorized(auth.UNAUTHORIZED))
+    }
+
+    else if (hasSufficientRole({ minRole })(req.user)) {
+      return next()
```
Comment (des-des) [inline src/middleware/permissions.js:80]: This still makes no sense to me.
This line here says the user is allowed to access the resource, but at this point it is possible that the user is trying to access a resource that they are not allowed, as some of the checks have not been made.

```js
if (userFailsFirstCheck) unautherised
if (userFailsSomeOtherCheck) unautherised
if (somethingElse) unautherised

// authorised!!
```

If you think I am wrong, then ill merge this tomorrow morning.

**Review by des-des — CHANGES_REQUESTED** [review]

File: src/middleware/permissions.js (hunk 6 above)
Comment (m4v15) [reply src/middleware/permissions.js:80]: Which checks haven't been made?

Comment (des-des) [reply src/middleware/permissions.js:80]: okay, I see what is going on now (https://github.com/foundersandcoders/open-tourism-platform/issues/134#issuecomment-330338072)

**Review by des-des — APPROVED** [review]

**Comment by des-des** [comment]
happy for you to merge

**Pull request merged**